flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/SplitRemoteConditionFromJoinRule.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.plan.rules.logical;
+
+import org.apache.flink.table.api.TableException;
+import org.apache.flink.table.planner.plan.nodes.logical.FlinkLogicalCalc;
+import org.apache.flink.table.planner.plan.nodes.logical.FlinkLogicalJoin;
+
+import org.apache.calcite.plan.RelOptRule;
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelOptUtil;
+import org.apache.calcite.rel.core.JoinRelType;
+import org.apache.calcite.rex.RexBuilder;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexProgram;
+import org.apache.calcite.rex.RexProgramBuilder;
+import org.apache.calcite.rex.RexUtil;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * Rule will split the {@link FlinkLogicalJoin} which contains Remote Functions in join condition
+ * into a {@link FlinkLogicalJoin} and a {@link FlinkLogicalCalc} with Remote Functions. Currently,
+ * only inner join is supported.
+ *
+ * <p>After this rule is applied, there will be no Remote Functions in the condition of the {@link
+ * FlinkLogicalJoin}.
+ */
+public final class SplitRemoteConditionFromJoinRule extends RelOptRule {
+    protected final RemoteCalcCallFinder callFinder;
+    protected final Optional<String> errorOnUnsplittableRemoteCall;
+
+    public SplitRemoteConditionFromJoinRule(
+            RemoteCalcCallFinder callFinder, Optional<String> errorOnUnsplittableRemoteCall) {
+        super(operand(FlinkLogicalJoin.class, none()), "SplitRemoteConditionFromJoinRule");
```

Comment:
Man point of migration to java is replacement of deprecated api
here `operand` and `none` are still deprecated